### PR TITLE
Implement receiver list as array instead of linked list

### PR DESCRIPTION
**Pull request by mizu-mochi**

### CODE REVIEW

File: contracts/distribution/AlphaReleaseRuleSelector.sol [hunk 1]
```diff
@@ -85,14 +95,12 @@ contract AlphaReleaseRuleSelector is Ownable, IAlphaReleaseRuleSelector {
...
     returns (IAlphaReceiver[] memory, uint256[] memory)
   {
-    IAlphaReceiver[] memory receivers = new IAlphaReceiver[](ruleCount);
-    uint256[] memory amounts = new uint256[](ruleCount);
-    address currentReceivers = receiverList[HEAD];
-    for (uint256 i = 0; i < ruleCount; i++) {
-      receivers[i] = IAlphaReceiver(currentReceivers);
-      IAlphaReleaseRule releaseRule = rules[currentReceivers];
+    IAlphaReceiver[] memory receivers = new IAlphaReceiver[](receiverList.length);
+    uint256[] memory amounts = new uint256[](receiverList.length);
```
Comment (n1punp) [inline contracts/distribution/AlphaReleaseRuleSelector.sol]: Use receiverCount  instead of receiverList.length? (lines 98,99,100)

**Review by n1punp — APPROVED** [review]

File: contracts/distribution/AlphaReleaseRuleSelector.sol (hunk 1 above)
Comment (mizu-mochi) [reply contracts/distribution/AlphaReleaseRuleSelector.sol]: Fixed

File: contracts/distribution/AlphaReleaseRuleSelector.sol [hunk 2]
```diff
@@ -53,23 +50,38 @@ contract AlphaReleaseRuleSelector is Ownable, IAlphaReleaseRuleSelector {
...
-    rules[address(_receiver)] = _rule;
+    ReceiverRule memory receiverRule = ReceiverRule(
+      _receiver,
+      _rule
+    );
+    receiverRuleList.push(receiverRule);
     emit AlphaReleaseRuleUpdated(address(_receiver), address(_rule));
   }
 
+  function removeAlphaReleaseRule(IAlphaReceiver _receiver)
```
Comment (false10) [inline contracts/distribution/AlphaReleaseRuleSelector.sol]: receive index of list instead.

File: contracts/distribution/AlphaReleaseRuleSelector.sol [hunk 3]
```diff
@@ -12,22 +13,22 @@ import "../interfaces/IAlphaReleaseRuleSelector.sol";
...
+    // release rule
+    IAlphaReleaseRule rule;
+  }
+
   /**
-   * @dev the number of all rules registered to the Alpha release rule selector
+   * @dev the list of receivers with rule
    */
-  uint256 public ruleCount;
+  ReceiverRule[] public receiverRuleList;
```
Comment (false10) [inline contracts/distribution/AlphaReleaseRuleSelector.sol:31]: Add view function to get the length of list.

File: contracts/distribution/AlphaReleaseRuleSelector.sol [hunk 4]
```diff
@@ -53,23 +50,38 @@ contract AlphaReleaseRuleSelector is Ownable, IAlphaReleaseRuleSelector {
...
+        break;
+      }
+    }
+  }
+
   /**
    * @dev get the release rule of the receiver
    * @param _receiver the receiver to get the release rule
    */
   function getReleaseRule(IAlphaReceiver _receiver) external view returns (IAlphaReleaseRule) {
```
Comment (false10) [inline contracts/distribution/AlphaReleaseRuleSelector.sol]: Remove this function.

**Review by false10 — CHANGES_REQUESTED** [review]

File: contracts/distribution/AlphaReleaseRuleSelector.sol (hunk 2 above)
Comment (mizu-mochi) [reply contracts/distribution/AlphaReleaseRuleSelector.sol]: Fixed

File: contracts/distribution/AlphaReleaseRuleSelector.sol (hunk 3 above)
Comment (mizu-mochi) [reply contracts/distribution/AlphaReleaseRuleSelector.sol:31]: Added view function

File: contracts/distribution/AlphaReleaseRuleSelector.sol (hunk 4 above)
Comment (mizu-mochi) [reply contracts/distribution/AlphaReleaseRuleSelector.sol]: Removed

**Review by n1punp — APPROVED** [review]

File: contracts/distribution/AlphaReleaseRuleSelector.sol [hunk 5]
```diff
@@ -53,23 +61,30 @@ contract AlphaReleaseRuleSelector is Ownable, IAlphaReleaseRuleSelector {
...
   }
 
+  function removeAlphaReleaseRule(uint256 _index)
+    external
+    onlyOwner
+  {
+    ReceiverRule storage removedReceiverRule = receiverRuleList[_index];
+    receiverRuleList[_index] = receiverRuleList[receiverRuleList.length.sub(1)];
+    receiverRuleList.pop();
+    emit AlphaReleaseRuleRemoved(_index, address(removedReceiverRule.receiver), address(removedReceiverRule.rule));
```
Comment (false10) [inline contracts/distribution/AlphaReleaseRuleSelector.sol:83]: Emit rule index would be great?

**Review by false10 — CHANGES_REQUESTED** [review]

File: contracts/distribution/AlphaReleaseRuleSelector.sol (hunk 5 above)
Comment (mizu-mochi) [reply contracts/distribution/AlphaReleaseRuleSelector.sol:83]: Removed index from event

**Review by false10 — APPROVED** [review]

**Pull request merged**